### PR TITLE
Turn on dual slope ToT in phase-2 pixels digitizer and clusterizer and update CPE payloads to match

### PR DESCRIPTION
**Pull request by OzAmram**
This pull request turns on the 'dual-slope' configuration of the phase 2 pixel electronics simulation, and changes the Q/ToT settings from 600e to 1500e and the threshhold from 1200e to 1000e. 

To match these changes GenErrors and Templates were made for the T21 and T22 geometries, and these are being activated as well. The 'generic reco only' configuration of the T22 geoemtry is turned off as well now that the necessary payloads exist. 

The performance of these new payloads in local reconstruction were tested and look good.

Tracking validation was also run with the new changes using 100 TTbar events without pileup for both the T21 and T22 geometries (workflows 29034 and 29434). They show only minor differences as expected. 

These tests are summarized in slides [here](https://indico.cern.ch/event/934808/contributions/3996590/attachments/2094091/3519320/phase2_dual_slope_payloads_and_testing.pdf). 

@mmusich @tsusa @emiglior @cmantill @mtosi @tvami @pmaksim1 @JanFSchulte @jalimena @perrotta @slava77

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test workflow 29034.0,29434.0

**Comment by kpedro88** [comment]
+upgrade
changes limited to Phase 2 workflows, arising from differences in the tracking
tracking performance is expected to be similar; see e.g. only minor fluctuations in general track pT and PF candidate pT:
all_OldVSNew_TTbar14TeV2026D49wf23234p0c_log10recoTracks_generalTracks__RECO_obj_pt_2
all_OldVSNew_TTbar14TeV2026D49wf23234p0c_log10recoPFCandidates_particleFlow__RECO_obj_pt

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py [hunk 1]
```diff
@@ -2,17 +2,17 @@
 import FWCore.ParameterSet.Config as cms
 
 PixelDigitizerAlgorithmCommon = cms.PSet(
-    ElectronPerAdc = cms.double(600.0),
+    ElectronPerAdc = cms.double(1500.0),
```
Comment (jpata) [inline SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py:5]: It was noticed by Slava that these numbers are repeated from https://github.com/cms-sw/cmssw/pull/31253/files#diff-4e4e386df550fa27f52ff680d9d0f183R40. Could you make them configurable form one location, if they are always intended to be the same?

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py [hunk 2]
```diff
@@ -2,17 +2,17 @@
...
     ReadoutNoiseInElec = cms.double(0.0),
-    ThresholdInElectrons_Barrel = cms.double(1200.0),
-    ThresholdInElectrons_Endcap = cms.double(1200.0),
+    ThresholdInElectrons_Barrel = cms.double(1000.0),
```
Comment (jpata) [inline SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py:7]: same as https://github.com/cms-sw/cmssw/pull/31253/files#r479145041

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py [hunk 3]
```diff
@@ -2,17 +2,17 @@
...
+    ThresholdInElectrons_Endcap = cms.double(1000.0),
     AddThresholdSmearing = cms.bool(False),
     ThresholdSmearing_Barrel = cms.double(0.0),
     ThresholdSmearing_Endcap = cms.double(0.0),
     HIPThresholdInElectrons_Barrel = cms.double(1.0e10), # very high value to avoid Over threshold bit
     HIPThresholdInElectrons_Endcap = cms.double(1.0e10), # very high value to avoid Over threshold bit
     NoiseInElectrons = cms.double(0.0),
-    Phase2ReadoutMode = cms.int32(-1), # Flag to decide Readout Mode :Digital(0) or Analog (linear TDR (-1), dual slope with slope parameters (+1,+2,+3,+4) with threshold subtraction
+    Phase2ReadoutMode = cms.int32(3), # Flag to decide Readout Mode :Digital(0) or Analog (linear TDR (-1), dual slope with slope parameters (+1,+2,+3,+4) with threshold subtraction
```
Comment (jpata) [inline SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py:15]: same as https://github.com/cms-sw/cmssw/pull/31253/files#r479145041

**Comment by jpata** [comment]
Thanks for this PR @OzAmram! Just for documentation, was this presented in any tracking meeting yet? If so, could you put a an indico link to the PR description as well? I think it would be good to add some references for the bugfix (before/after?).

I understand that even a minor change to tracking will result in changes to most of the downstream reconstruction values, hence we have `Reco comparison results: 5156 differences found in the comparisons`, nothing to be done here I guess.

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py (hunk 1 above)
Comment (mmusich) [reply SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py:5]: I'd rather take the clusterizer numbers from the digitizer configuration than the other way around, as for now the clusterizer is taking the simDigis in input as we lack a phase-2 unpacker/packer setup (so the clusterizer needs to stay in synch with the digitizer configuration).

**Comment by kpedro88** [comment]
@jpata two sets of slides are linked in the PR description

File: SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py (hunk 1 above)
Comment (OzAmram) [reply SimTracker/SiPhase2Digitizer/python/phase2TrackerDigitizer_cfi.py:5]: Ok that makes sense. I just pushed a new commit that has the clusterizer config import the values from the digitizer. Let me know if there is some other way people want this to be implemented.

**Comment by OzAmram** [comment]
The material for the PR was not presented in any meetings yet. I did link to material in the PR description that was circulated around before making the PR so it didn't seem necessary and can serve as documentation for posterity. I can make a presentation in a meeting about this if one of conveners wish

**Comment by OzAmram** [comment]
The new commit has the clusterizer import its configuration values from the digitizer to address a code comment, it should be functionally the same

File: RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py [hunk 4]
```diff
@@ -30,14 +30,15 @@
...
 # Need these until phase2 pixel templates are used
 from Configuration.Eras.Modifier_phase2_tracker_cff import phase2_tracker
+from SimTracker.SiPhase2Digitizer.phase2TrackerDigitizer_cfi import PixelDigitizerAlgorithmCommon
 phase2_tracker.toModify(siPixelClusters, # FIXME
   src = 'simSiPixelDigis:Pixel',
   MissCalibrate = False,
   Phase2Calibration = True,
-  Phase2ReadoutMode = -1, # Flag to decide Readout Mode : linear TDR (-1), dual slope with slope parameters (+1,+2,+3,+4 ...) with threshold subtraction
-  Phase2DigiBaseline = 1200.,
+  Phase2ReadoutMode = 3, # Flag to decide Readout Mode : linear TDR (-1), dual slope with slope parameters (+1,+2,+3,+4 ...) with threshold subtraction
```
Comment (kpedro88) [inline RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py:38]: should this also be taken from `PixelDigitizerAlgorithmCommon`?

Comment (OzAmram) [reply RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py:38]: Ah you're right. I'll update

**Comment by civanch** [comment]
please test

File: RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py [hunk 5]
```diff
@@ -30,14 +30,15 @@
...
 # Need these until phase2 pixel templates are used
 from Configuration.Eras.Modifier_phase2_tracker_cff import phase2_tracker
+from SimTracker.SiPhase2Digitizer.phase2TrackerDigitizer_cfi import PixelDigitizerAlgorithmCommon
 phase2_tracker.toModify(siPixelClusters, # FIXME
   src = 'simSiPixelDigis:Pixel',
   MissCalibrate = False,
   Phase2Calibration = True,
-  Phase2ReadoutMode = -1, # Flag to decide Readout Mode : linear TDR (-1), dual slope with slope parameters (+1,+2,+3,+4 ...) with threshold subtraction
-  Phase2DigiBaseline = 1200.,
+  Phase2ReadoutMode = PixelDigitizerAlgorithmCommon.Phase2ReadoutMode, # Flag to decide Readout Mode : linear TDR (-1), dual slope with slope parameters (+1,+2,+3,+4 ...) with threshold subtraction
```
Comment (mmusich) [inline RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py:38]: to be extremely picky (an inline with the current good practices as seen in recent RECO PRs) one could drop the type specification and use  `PixelDigitizerAlgorithmCommon.Phase2ReadoutMode.value()`

**Comment by jpata** [comment]
> @jpata two sets of slides are linked in the PR description

Thanks, I did see the slides, but was asking explicitly about a POG/DPG presentation with the corresponding indico link. Private cernbox links may be somewhat less durable.

@OzAmram
> The material for the PR was not presented in any meetings yet. I did link to material in the PR description that was circulated around before making the PR so it didn't seem necessary and can serve as documentation for posterity. I can make a presentation in a meeting about this if one of conveners wish

Can you ping the right convener here please, just to make sure?

**Comment by mmusich** [comment]
@jpata

>Can you ping the right convener here please, just to make sure?

We're all already copied... check the PR description. 

@OzAmram would you like to show the slides tomorrow at the DPG meeting for future reference?

**Comment by jpata** [comment]
> @jpata
> 
> > Can you ping the right convener here please, just to make sure?
> 
> We're all already copied... check the PR description.
> 
> @OzAmram would you like to show the slides tomorrow at the DPG meeting for future reference?

Thanks @mmusich - trying to memorize better who is in which role. Just to clarify, from the reco point of view a formal presentation is not explicitly needed if the tracking group is fine with it, confirmed on this PR.

File: RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py (hunk 5 above)
Comment (makortel) [reply RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py:38]: It's not only about the type specification. Configuration like
```python
phase2_tracker.toModify(siPixelClusters,
    ...
    Phase2ReadoutMode = PixelDigitizerAlgorithmCommon.Phase2ReadoutMode
    ...
)
```
would make `siPixelClusters.Phase2ReadoutMode` to be exactly the same `cms.int32` object as `PixelDigitizerAlgorithmCommon.Phase2ReadoutMode`, which means that if anyone would change the value of `PixelDigitizerAlgorithmCommon.Phase2ReadoutMode` later in the python program execution without "specifying the type", also `siPixelClusters.Phase2ReadoutMode` would change automatically (being the same object). Example code:
```python
from Configuration.Eras.Modifier_phase2_tracker_cff import phase2_tracker
process = cms.Process("TEST", phase2_tracker)
process.load("RecoLocalTracker.SiPixelClusterizer.SiPixelClusterizer_cfi")

from SimTracker.SiPhase2Digitizer.phase2TrackerDigitizer_cfi import PixelDigitizerAlgorithmCommon
PixelDigitizerAlgorithmCommon.Phase2ReadoutMode = 5
# now siPixelClusters.Phase2ReadoutMode == 5
```
Usually this kind of "long distance interaction" can lead to very confusing behavior, which is why core recommends to always `clone()`/`copy()`/`value()` the value of `cms.<type>` when assigning from an existing configuration object to a new one.

**Comment by OzAmram** [comment]
Ok I pushed a new commit that uses the value() for the importing of settings from the configuration object

@mmusich I don't think I can make the meeting tomorrow, but maybe I can just upload the slides to the agenda for posterity?

**Comment by mmusich** [comment]
> @mmusich I don't think I can make the meeting tomorrow, but maybe I can just upload the slides to the agenda for posterity?

Thanks Oz, that works for me. Slot is here: https://indico.cern.ch/event/934808/#59-updated-phase-2-cpe-payload

**Comment by OzAmram** [comment]
Thanks. I uploaded slides that summarize both the residuals and tracking validation tests. PR description is now updated to refer to these slides.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
+reconstruction

- updates phase 2 tracking with the dual-slope configuration, includes also a bugfix in `PixelThresholdClusterizer.cc`
- there are a large number of minute differences from changes to tracking, but the overall tracking performance has not changed, as validated on workflows 29034.0 and 29434.0, slides have been attached to a recent Tracker DPG agenda
- the configuration values in `SiPixelClusterizer_cfi.py` are now loaded from `phase2TrackerDigitizer_cfi.py`

**Comment by tlampen** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by mmusich** [comment]
@cms-sw/pdmv-l2 @cms-sw/upgrade-l2 any further comment?

**Comment by mmusich** [comment]
@kpedro88  @chayanit any further comment on this PR? Thank you.

**Comment by kpedro88** [comment]
+upgrade

**Comment by chayanit** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**